Userland/Libraries/LibSQL/TreeNode.cpp

```diff
@@ -0,0 +1,284 @@
+    for (auto ix = 0u; ix < size(); ix++) {
+        auto& entry = m_entries[ix];
+        serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[ix]);
+        entry.serialize(buffer);
+    }
+    serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[size()]);
+}
+
+void TreeNode::just_insert(Key const& key, TreeNode *right)
+{
+    dbgln("[{}] just_insert({}, right = {})",
+        m_pointer, (String) key, (right) ? right->m_pointer : -1);
+    VERIFY(size() < max_entries());
+    for (auto ix = 0u; ix < size(); ix++) {
+        if (key < m_entries[ix]) {
+            m_entries.insert(ix, key);
+            if (right) {
+                m_down_pointers.insert(ix+1, right->pointer());
+                m_down_nodes.insert(ix + 1, right);
+            }
+            tree()->add_to_write_ahead_log(*this);
+            return;
+        }
+    }
+    m_entries.append(key);
+    if (right) {
+        m_down_pointers.append(right->pointer());
+        m_down_nodes.append(right);
+    }
+    tree()->add_to_write_ahead_log(*this);
+}
+
+void TreeNode::split_and_insert(Key const& key, TreeNode *right)
+{
+    dbgln("[{}] split_and_insert({}, right = {})",
+        m_pointer, (String) key, (right) ? right->m_pointer : -1);
+    if (!m_up) {
+        // Make new m_up. This is the new root node.
+        m_up = m_tree->new_root();
+    }
+
+    // Find median:
+    Key median;
+    dbgln("[{}] Splitting node. size: {} down_nodes: {}",
+        m_pointer, size(), m_down_nodes.size());
+    auto left = (!m_down_nodes.is_empty()) ? m_down_nodes[max_entries() / 2 + 2] : nullptr;
+    auto new_node = adopt_ref(*new TreeNode(*m_tree, m_up, left));
+    
+    dbgln("[{}] New node: {} up: {}", m_pointer, new_node->m_pointer, m_up->m_pointer);
+
+    auto move_right = max_entries() / 2;      // Number of entries to move to the new, right, node
+    if (m_entries[max_entries() / 2] < key) { // New entry belongs in new, right, node.
+        median = m_entries[max_entries() / 2];
+        move_right--;
+        new_node->just_insert(key, right);
+    } else if (m_entries[max_entries() / 2 - 1] > key) { // New entry belongs in current node,
+        // which will be the new left node.
+        median = m_entries.take(max_entries() / 2 - 1);
+        just_insert(key, right);
+    } else { // New entry is the median.
+        median = key;
+    }
+    for (size_t i = 0; i < move_right; i++) {
+        if (!m_down_nodes.is_empty()) {
+            new_node->just_insert(m_entries.take_last(), m_down_nodes.take_last());
+        } else {
+            new_node->just_insert(m_entries.take_last());
+        }
+    }
+    tree()->add_to_write_ahead_log(*this);
+    m_up->insert(median, new_node);
+}
+
+void TreeNode::dump() {
+    StringBuilder builder;
+    builder.appendff("[{}] size {}", pointer(), (int) size());
+    if (is_leaf()) {
+        builder.append(" is leaf");
+    } else {
+        builder.appendff(" down nodes {}", m_down_nodes.size());
+    }
```

Comment:
when all paths of the comparison are just one line, we usually omit the curly braces

you might have to fix it on earlyer lines too, this is the first time that I conciously see it

Reply:
30 years of habit don't go away that easily, young padawan :-)

I know there's discussions bordering on holy wars on what's best here; I've always used braces but it's not worth the aggravation. I'll remove them.